executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,546 @@
+                    set.insert(*v);
+                }
+            }),
+            InterpreterAction::Assertion(lhs, rhs, _) => {
+                lhs.elems.iter().for_each(|e| {
+                    if let RPNExpressionElem::Symbol(idx) = e {
+                        set.insert(*idx);
+                    }
+                });
+                rhs.elems.iter().for_each(|e| {
+                    if let RPNExpressionElem::Symbol(idx) = e {
+                        set.insert(*idx);
+                    }
+                });
+            }
+            _ => {}
+        }
+        set
+    }
+}
+
+/// Helper struct to map variables to unique indices, so they can be kept in
+/// sequential memory and quickly refered to during execution.
+pub struct VariableMapper {
+    var_idx: HashMap<Variable, usize>,
+    count: usize,
+}
+
+impl VariableMapper {
+    pub fn new() -> Self {
+        Self {
+            var_idx: HashMap::new(),
+            count: 0,
+        }
+    }
+
+    pub fn var_count(&self) -> usize {
+        self.count
+    }
+
+    pub fn map_var(&mut self, var: &Variable) -> usize {
+        let idx = *self.var_idx.entry(var.clone()).or_insert_with(|| {
+            self.count += 1;
+            self.count - 1
+        });
+        idx
+    }
+
+    // reserve a new variable index
+    pub fn reserve_idx(&mut self) -> usize {
+        let idx = self.count;
+        self.count += 1;
+        idx
+    }
+
+    // get the index of a variable if it was previously mapped
+    pub fn get_var(&mut self, var: &Variable) -> Option<usize> {
+        self.var_idx.get(var).copied()
+    }
+
+    pub fn map_expr_to_rpn<T: FieldElement>(
+        &mut self,
+        expr: &SymbolicExpression<T, Variable>,
+    ) -> RPNExpression<T, usize> {
+        RPNExpression::from(&expr.map_variables(&mut |var| self.map_var(var)))
+    }
+}
+
+/// An expression in Reverse Polish Notation.
+pub struct RPNExpression<T: FieldElement, S> {
+    pub elems: Vec<RPNExpressionElem<T, S>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum RPNExpressionElem<T: FieldElement, S> {
+    Concrete(T),
+    Symbol(S),
+    BinaryOperation(BinaryOperator),
+    UnaryOperation(UnaryOperator),
+    BitOperation(BitOperator, T::Integer),
+}
+
+impl<T: FieldElement, S: Clone> From<&SymbolicExpression<T, S>> for RPNExpression<T, S> {
+    fn from(expr: &SymbolicExpression<T, S>) -> Self {
+        fn from_inner<T: FieldElement, S: Clone>(
+            expr: &SymbolicExpression<T, S>,
+            elems: &mut Vec<RPNExpressionElem<T, S>>,
+        ) {
+            match expr {
+                SymbolicExpression::Concrete(n) => {
+                    elems.push(RPNExpressionElem::Concrete(*n));
+                }
+                SymbolicExpression::Symbol(s, _) => {
+                    elems.push(RPNExpressionElem::Symbol(s.clone()));
+                }
+                SymbolicExpression::BinaryOperation(lhs, op, rhs, _) => {
+                    from_inner(lhs, elems);
```

Comment:
Returning an iterator instead of pushing to a `&mut` does not work (or would require a `Box<dyn Iterator`) because of these recursive calls, right?

Reply:
not sure, i think this way is clearer, we're building the vec we're using anyways and its internal